kratos/utilities/model_part_graph_utilities.cpp

```diff
@@ -0,0 +1,262 @@
+        if(aux_list.size() != nnodes)
+            aux_list.resize(nnodes);
+
+        for(ModelPartGraphUtilities::IndexType i=0; i<nnodes; ++i)
+        {
+            aux_list[i] = r_geom[i].Id()-1;
+        }
+        Agraph->AddEntries(aux_list);
+    });
+    Agraph->Finalize();
+
+    return Agraph;
+}
+
+std::pair<DenseVector<ModelPartGraphUtilities::IndexType>, DenseVector<ModelPartGraphUtilities::IndexType>> ModelPartGraphUtilities::ComputeCSRGraph(const ModelPart& rModelPart)
+{
+    std::pair<DenseVector<ModelPartGraphUtilities::IndexType>, DenseVector<ModelPartGraphUtilities::IndexType>> data;
+    auto& rRowIndices = data.first;
+    auto& rColIndices = data.second;
+    const auto graph = ComputeGraph(rModelPart);
+    graph->ExportCSRArrays(rRowIndices,rColIndices);
+    return data;
+}
+
+std::pair<ModelPartGraphUtilities::IndexType, DenseVector<double>> ModelPartGraphUtilities::ComputeConnectedComponents(
+            const ModelPart::NodesContainerType& rNodes,
+            const DenseVector<ModelPartGraphUtilities::IndexType>& rRowIndices,
+            const DenseVector<ModelPartGraphUtilities::IndexType>& rColIndices
+        )
+{
+    const ModelPartGraphUtilities::IndexType nindices = rNodes.size();
+
+    std::unordered_map<ModelPartGraphUtilities::IndexType, int> visited(nindices);
+
+    //set -1 as a special "not visited" value in visited - nodes will be marked as visited when visited[i]!=-1
+    //note that we subtract -1 from the Id of the nodes to have all the indices to start in 0
+    for(const auto& rNode : rNodes)
+        visited[rNode.Id()-1] = -1;
+
+    ModelPartGraphUtilities::IndexType color=0;
+    for(ModelPartGraphUtilities::IndexType i=0; i<nindices; ++i)
+    {
+        const ModelPartGraphUtilities::IndexType gid = (rNodes.begin()+i)->Id()-1;
+        if(visited[gid] == -1 && (rRowIndices[gid+1]-rRowIndices[gid])!=0)
+        {
+            ModelPartGraphUtilities::IndexType root=gid;
+            BreadthFirstSearch(root, color, rRowIndices, rColIndices, visited);
+            color += 1;
+        }
+    }
+
+    auto number_of_colors = color;
+
+    //prepare output so that is can be directly mapped to nodes by VariableUtils
+    //even in the case in which nodes are not numbered consecutively
+    DenseVector<double> colors(nindices);
+    ModelPartGraphUtilities::IndexType counter=0;
+    for(const auto& rNode : rNodes)
+        colors[counter++] = visited[rNode.Id()-1];
+
+    return std::pair{number_of_colors, colors};
+}
+
+std::pair<ModelPartGraphUtilities::IndexType, DenseVector<double>> ModelPartGraphUtilities::ComputeConnectedComponents_ActiveNodesCheck(
+            const ModelPart::NodesContainerType& rNodes,
+            const DenseVector<ModelPartGraphUtilities::IndexType>& rRowIndices,
+            const DenseVector<ModelPartGraphUtilities::IndexType>& rColIndices,
+            const std::vector<bool>& active_nodes_list
```

Comment:
```suggestion
    const ModelPart::NodesContainerType& rNodes,
    const DenseVector<ModelPartGraphUtilities::IndexType>& rRowIndices,
    const DenseVector<ModelPartGraphUtilities::IndexType>& rColIndices,
    const std::vector<bool>& active_nodes_list
```
BTW: Should be rActiveNodesList